jib-core/src/main/java/com/google/cloud/tools/jib/filesystem/DirectoryWalker.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.filesystem;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.stream.Stream;
+
+/** Recursively applies a function to each file in a directory. */
+public class DirectoryWalker {
```

Comment:
I feel like we're re-implementing something from java.nio here

Reply:
Yea, this is mostly just `Files#walk`, but is able to account for `IOException`s in the lambda so we don't have to have the Unchecked to checked conversion everywhere.

Reply:
Hah, that just sounds like maybe we should take another look at it?

Reply:
Hmm, I've looked into how others handle exceptions in Streams and the only solution I've found so far is to throw an unchecked exception. Perhaps we should not use Files.walk?

Example:
https://www.oreilly.com/ideas/handling-checked-exceptions-in-java-streams

Reply:
Yeah, I don't know why, but their solution just seems so hacky. Creating this whole uncheckedIOException and then catching that unchecked exception and throwing the cause, it gives me a gross feeling. But I guess it is what it is.

Reply:
Yea, it is definitely not an elegant way to do it. Another way is to do what Guava does for `ListenableFuture`s, which I believe is wrap all exceptions in an `ExecutionException` to be caught and handled by the user.